bookshelf/routes/web.php

```diff
@@ -0,0 +1,199 @@
+| Application Routes
+|--------------------------------------------------------------------------
+|
+| Here is where you can register all of the routes for an application.
+| It is a breeze. Simply tell Lumen the URIs it should respond to
+| and give it the Closure to call when that URI is requested.
+|
+*/
+
+$router->get('/', function (Request $request) use ($collection) {
+    $pageSize = getenv('BOOKSHELF_PAGE_SIZE') ?: 10;
+    $query = $collection->limit($pageSize)->orderBy('title');
+    if ($token = $request->query->get('page_token')) {
+        $lastBook = $collection->document($token)->snapshot();
+        $query = $query->startAfter($lastBook);
+    }
+
+    return view('list', [
+        'books' => $query->documents(),
+        'pageSize' => $pageSize,
+    ]);
+});
+
+
+// [START add]
+$router->get('/books/add', function () {
+    return view('form', [
+        'action' => 'Add',
+        'book' => [],
+    ]);
+});
+
+$router->post('/books/add', function (Request $request) use ($collection, $gcsBucket) {
+    $files = $request->files;
+    $bookData = $request->request->all();
+    $image = $files->get('image');
+    if ($image && $image->isValid()) {
+        $f = fopen($image->getRealPath(), 'r');
+        $object = $gcsBucket->upload($f, [
+            'metadata' => ['contentType' => $image->getMimeType()],
+            'predefinedAcl' => 'publicRead',
+        ]);
+        $bookData['image_url'] = $object->info()['mediaLink'];
+    }
+
+    // Create the book
+    $bookRef = $collection->newDocument();
+    $bookRef->set($bookData);
+
+    return redirect("/books/" . $bookRef->id());
+});
+// [END add]
+
+// [START show]
+$router->get('/books/{bookId}', function ($bookId) use ($collection) {
+    # [START firestore_client_get_book]
+    $bookRef = $collection->document($bookId);
+    $snapshot = $bookRef->snapshot();
+    # [END firestore_client_get_book]
+
+    if (!$snapshot->exists()) {
+        return new Response('', Response::HTTP_NOT_FOUND);
+    }
+
+    return view('show', ['book' => $snapshot]);
+});
+// [END show]
+
+// [START edit]
+$router->get('/books/{bookId}/edit', function ($bookId) use ($collection) {
+    $bookRef = $collection->document($bookId);
+    $snapshot = $bookRef->snapshot();
+
+    if (!$snapshot->exists()) {
+        return new Response('', Response::HTTP_NOT_FOUND);
+    }
+
+    return view('form', [
+        'action' => 'Edit',
+        'book' => $snapshot,
+    ]);
+});
+
+$router->post('/books/{bookId}/edit', function (Request $request, $bookId) use ($collection, $gcsBucket) {
+    $bookRef = $collection->document($bookId);
+    $snapshot = $bookRef->snapshot();
+
+    if (!$snapshot->exists()) {
+        return new Response('', Response::HTTP_NOT_FOUND);
+    }
+
+    // Get book data from the request object
+    $bookData = $request->request->all();
+    $bookData['id'] = $bookId;
+
+    // [START add_image]
+    $files = $request->files;
```

Comment:
What do you think about turning this into a helper since the same logic lives in the `/books/add` action?

Reply:
I am not opposed to adding a helper function but I thought in this case, not being DRY would make the code easier to follow.